app/src/main/java/com/practicum/playlistmaker/db/data/PlaylistRepositoryImpl.kt

```diff
@@ -23,13 +29,107 @@ class PlaylistRepositoryImpl(
         emit(convertFromPlaylistEntity(playlists))
     }
 
+    override suspend fun addTrackToPL(track: Track) {
+        appDatabase.pLTrackDao().insertTrack(playlistDbConvertor.mapTrack(track))
+    }
+
+    override fun getTracksPL(id: List<Int>): Flow<List<Track>> = flow {
+        emit(tracksList(id))
+    }
+
+    override suspend fun getCurrentPL(id: Long): Playlist {
+        return playlistDbConvertor.map(appDatabase.playlistDao().getCurrentPL(id))
+    }
+
+    override fun deleteTrack(idPL: Long, idTrack: Int): Flow<List<Track>> = flow {
+        val pl = appDatabase.playlistDao().getCurrentPL(idPL)
+        var idList = playlistDbConvertor.mapToList(pl.tracks_id) as MutableList
+        for (item in idList) {
+            if (idTrack == item) {
+                idList.remove(idTrack)
+                break
+            }
+        }
+        if (!idList.isEmpty()) {
+            val newTracksID = playlistDbConvertor.mapToJson(idList)
+            val count = pl.tracksCount - 1
+            val newPL = Playlist(
+                pl.plID,
+                pl.playlistName,
+                pl.playlistdescription,
+                pl.uri,
+                newTracksID,
+                count
+            )
+            addToPlaylist(newPL)
+        } else {
+            val newPL = Playlist(
+                pl.plID,
+                pl.playlistName,
+                pl.playlistdescription,
+                pl.uri,
+                "",
+                0
+            )
+            addToPlaylist(newPL)
+        }
+
+        if (!checkingPlaylists(idTrack)) {
+            val currentTack = appDatabase.pLTrackDao().getCurrentTrack(idTrack)
+            appDatabase.pLTrackDao().deleteTrack(currentTack)
+        }
+        emit(tracksList(idList))
+
+    }
+
+
     private fun convertFromPlaylistEntity(playlists: List<PlaylistEntity>): List<Playlist> {
         return playlists.map { playlists -> playlistDbConvertor.map(playlists) }
     }
 
-    override suspend fun addTrackToPL(track: Track) {
-        appDatabase.pLTrackDao().insertTrack(playlistDbConvertor.mapTrack(track))
+    private suspend fun tracksList(id: List<Int>): List<Track> {
+        val currentPLTracks = ArrayList<Track>()
+        val allTracks = appDatabase.pLTrackDao().getTracks()
+        for (item in id) {
+            for (track in allTracks)
+                if (item == track.trackId) {
+                    currentPLTracks.add(trackDbConvertor.map(track))
+                }
+        }
+        return currentPLTracks
+    }
+
+    private suspend fun checkingPlaylists(id: Int): Boolean {
+        var isInPL = false
+        val plList = appDatabase.playlistDao().getPlaylists()
+        for (item in plList) {
+            if (item.tracks_id != "") {
+                val idTracks = playlistDbConvertor.mapToList(item.tracks_id) as MutableList
```

Comment:
Все касты лучше делать безопасными, чтобы не было крешей:
```suggestion
                val idTracks = playlistDbConvertor.mapToList(item.tracks_id) as? MutableList
```